base/asyncmap.jl

```diff
@@ -72,14 +73,32 @@ function next(itr::AsyncCollector, state::AsyncCollectorState)
 
     # Execute function call and save result asynchronously
     @async begin
-        itr.results[i] = itr.f(args...)
-        state.active_count -= 1
-        notify(state.task_done, nothing)
+        try
+            itr.results[i] = itr.f(args...)
+        catch e
+            try
+                if isa(itr.on_error, Function)
+                    itr.results[i] = itr.on_error(e)
+                else
+                    rethrow(e)
+                end
+            catch e2
+                state.in_error = true
+                notify(state.task_done, e2; error=true)
+
+                # The "notify" above raises an exception if "next" is waiting for tasks to finish.
+                # If the calling task is waiting on sync_end(), the rethrow() below will be captured
+                # by it.
+                rethrow(e2)
+            end
+        finally
+            state.active_count -= 1
+            notify(state.task_done, nothing)
+        end
```

Comment:
Ok, I think I get it...

If the mapped function `f` throws an error, then `notify(state.task_done, e2; error=true)` is intended to stop the iteration in the main task as soon as possible when there is an error (instead of waiting until `sync_end()`).

But, there is a race condition.
If the main task happens not to be waiting at the top of `next` when the error is thrown, it will not see the error.
This is handled by the `in_error` flag, because the next time the main iterator task calls `done` it will see the flag and call `sync_end()` immediately. `sync_end()` will then throw the error that caused the worker task to die at line 92.

It doesn't seem quite right to set `state.done = true` when `state.in_error == true` in `done` because the iterator did not actually finish.

What if `done` was rewritten like this:
```julia
function done(itr::AsyncCollector, state::AsyncCollectorState)
    if state.in_error
        sync_end()
        @assert false "unreachable" 
    end
    if !state.done && done(itr.enumerator, state.enum_state)
        state.done = true
        sync_end()
    end
    return state.done
end
```

I think that this more clearly differentiates between the done and error states. 

...now that I think about it some more, I think I would remove the `notify` on line 87.
This notify only has any effect if `next` is waiting in `wait(state)` at line 68.
`wait(state)` will be woken anyway by the `notify` on line 96.
The only difference is that the `notify` on line 96 does not have `error=true`. This means that the main iterator task will start one more `@async` tasks before getting back to `done` and discovering the error.
I would argue that the small efficiency gain of stopping a little earlier is not justified by the extra complexity of the extra `notify` on line 87. (i.e. it took a bit of figuring-out to figure out what was going on.

e.g.

```julia
            catch e2
                state.in_error = true
                # The task will die and the error will be rethrown by sync_end() in done().
                rethrow(e2)
            end
```